src/goto-instrument/contracts/contracts.cpp

```diff
@@ -294,6 +294,12 @@ void code_contractst::check_apply_loop_contracts(
   // Generate havocing code for assignment targets.
   havoc_assigns_targetst havoc_gen(to_havoc, ns);
   havoc_gen.append_full_havoc_code(loop_head_location, pre_loop_head_instrs);
+  // Add loop number to the havoc_code
+  for(auto &havoc_instr : pre_loop_head_instrs.instructions)
+  {
+    if(is_loop_havoc(havoc_instr))
+      havoc_instr.loop_number = loop_end->loop_number;
+  }
```

Comment:
After loop-contracts transformation, there is no loop and hence no loop number. But in the counter-example-guided synthesis we need locate which loop cause the violation so that we synthesize loop invariant for that loop. My idea is that record original loop number in some instrumented instructions to allow us know which loop the instruction transformed from.
But you are right that using `loop_number` here is ambiguous. I will use a new variable `original_loop_number` with `optionalt<unsigned>` type.

Reply:
As my reply to Daniel said, I will use a map to store the information about original loop numbers in `code_contractst` instead of using `original_loop_number`.